server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.snapshots;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ResultDeduplicator;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.snapshots.IndexShardSnapshotStatus;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Tracks progress of shard snapshots during shutdown, on this single data node. Periodically reports progress via logging, the interval for
+ * which see {@link #SNAPSHOT_PROGRESS_DURING_SHUTDOWN_INTERVAL_TIME_SETTING}.
+ */
+public class SnapshotShutdownProgressTracker {
+    /**
+     * Runnable that logs shard snapshot progress.
+     */
+    private class ProgressLogger implements Runnable {
+        @Override
+        public void run() {
+            logProgressReport();
+        }
+    }
+
+    /** How frequently shard snapshot progress is logged after receiving local node shutdown metadata. */
+    public static final Setting<TimeValue> SNAPSHOT_PROGRESS_DURING_SHUTDOWN_INTERVAL_TIME_SETTING = Setting.positiveTimeSetting(
```

Comment:
WDYT about making this setting dynamic? I don't think that would add much complexity.